yew-macro/src/html_tree/html_tag/mod.rs

```diff
@@ -195,6 +210,95 @@ impl ToTokens for HtmlTag {
     }
 }
 
+struct DynamicName {
+    at: Token![@],
+    expr: Option<Block>,
+}
+
+impl Peek<'_, ()> for DynamicName {
+    fn peek(cursor: Cursor) -> Option<((), Cursor)> {
+        let (punct, cursor) = cursor.punct()?;
+        (punct.as_char() == '@').as_option()?;
+
+        // move cursor past block if there is one
+        let cursor = cursor
+            .group(Delimiter::Brace)
+            .map(|(_, _, cursor)| cursor)
+            .unwrap_or(cursor);
+
+        Some(((), cursor))
+    }
+}
+
+impl Parse for DynamicName {
+    fn parse(input: ParseStream) -> ParseResult<Self> {
+        let at = input.parse()?;
+        let expr = if input.cursor().group(Delimiter::Brace).is_some() {
+            Some(input.parse()?)
+        } else {
+            None
+        };
+
+        Ok(Self { at, expr })
+    }
+}
+
+impl ToTokens for DynamicName {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        let Self { at, expr } = self;
+        tokens.extend(quote! {#at#expr});
+    }
+}
+
+#[derive(PartialEq)]
+enum TagKey {
```

Comment:
I felt it made sense to add a new type here.
It's possible to remove this type entirely and have the `peek` methods return  `TagName` directly but then it would have to implement `PartialEq` or have some sort of special comparison method to determine if a particular tag closes the current opening tag.